suite/volume/volume_setup.go

```diff
@@ -0,0 +1,320 @@
+			contactsStartTimes[contractAddr] = float64(minInt64Slice(startTimesAcrossNodes))
+			return nil
+		})
+	}
+	err := g.Wait()
+	if err != nil {
+		return nil, err
+	}
+	log.Debug().Interface("Round start times", contactsStartTimes).Send()
+	return contactsStartTimes, nil
+}
+
+func (vt *FluxTest) getAllRoundsSubmissionsEvents(ctx context.Context, roundID int, submissions int, submissionVal *big.Int) (map[string]float64, error) {
+	g, gctx := errgroup.WithContext(ctx)
+
+	mu := &sync.Mutex{}
+	endTimes := make(map[string]float64)
+	for _, fi := range *vt.FluxInstances {
+		fi := fi
+		g.Go(func() error {
+			endTime, err := fi.FilterRoundSubmissions(gctx, submissions, submissionVal, roundID)
+			if err != nil {
+				return err
+			}
+			mu.Lock()
+			defer mu.Unlock()
+			endTimes[fi.Address()] = float64(endTime)
+			return nil
+		})
+	}
+	err := g.Wait()
+	if err != nil {
+		return nil, err
+	}
+	return endTimes, nil
+}
+
+// roundsMetrics gets start times from runs via API, awaits submissions for each contract and get round end times
+func (vt *FluxTest) roundsMetrics(roundID int, submissions int, submissionVal *big.Int) error {
+	startTimes, err := vt.roundsStartTimes()
+	if err != nil {
+		return err
+	}
+	endTimes, err := vt.getAllRoundsSubmissionsEvents(context.Background(), roundID, submissions, submissionVal)
+	if err != nil {
+		return err
+	}
+	for contract := range startTimes {
+		duration := endTimes[contract] - startTimes[contract]
+		vt.roundsDurationData = append(vt.roundsDurationData, duration)
+		log.Info().Str("Contract", contract).Float64("Round duration ms", duration).Send()
+	}
+	return nil
+}
+
+// checkRoundDataOnChain check that ContractData about last round is correct
+func (vt *FluxTest) checkRoundDataOnChain(roundID int, newVal int) error {
+	if err := retry.Do(func() error {
+		finished, err := vt.checkContractData(roundID, newVal)
+		if err != nil {
+			return err
+		}
+		if !finished {
+			return errors.New("round is not finished")
+		}
+		return nil
+	}, vt.OnChainCheckAttemptsOpt); err != nil {
+		return errors.Wrap(err, "round is not fully finished on chain")
+	}
+	return nil
+}
+
+func (vt *FluxTest) checkContractData(roundID int, newVal int) (bool, error) {
+	log.Debug().Int("Round ID", roundID).Msg("Checking round completion on chain")
+	var rounds []*contracts.FluxAggregatorData
+	for _, flux := range *vt.FluxInstances {
+		cd, err := flux.GetContractData(context.Background())
+		if err != nil {
+			return false, err
+		}
+		rounds = append(rounds, cd)
+	}
+	vt.debugRoundTuple(rounds)
+	for _, r := range rounds {
+		if r.LatestRoundData.RoundId.Int64() != int64(roundID) || r.LatestRoundData.Answer.Int64() != int64(newVal) {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func (vt *FluxTest) debugRoundTuple(rounds []*contracts.FluxAggregatorData) {
+	roundValsArr := make([]roundVals, 0)
+	for _, r := range rounds {
+		if r == nil {
+			continue
+		}
+		roundValsArr = append(roundValsArr, roundVals{
```

Comment:
Do we really need a struct for this? Can't it just output the round ID & val without creating a struct?